stan/math/rev/mat/functor/coupled_ode_system.hpp

```diff
@@ -0,0 +1,241 @@
+       * coupled ODE system state with respect to time evaluated at the
+       * specified state and specified time.
+       *
+       * @param[in]  z the current state of the coupled ode system,
+       * of size <code>size()</code>.
+       * @param[in, out] dz_dt populate with the derivatives of the
+       * coupled system evaluated at the specified state and time.
+       * @param[in] t time.
+       * @throw exception if the base system does not return a
+       * derivative vector of the same size as the state vector.
+       *
+       * y is the base ODE system state
+       *
+       */
+      inline void
+      operator()(const std::vector<double>& z,
+                 std::vector<double>& dz_dt,
+                 double t) const {
+        rhs_sens(y0_, theta_, z, dz_dt, t);
+      }
+
+      /**
+       * Returns the size of the coupled system.
+       *
+       * @return size of the coupled system.
+       */
+      size_t size() const {
+        return size_;
+      }
+
+      /**
+       * Returns the initial state of the coupled system.
+       *
+       * In case the initials are varying, these get assigned the
+       * identity matrix as initial.
+       *
+       * @return the initial condition of the coupled system.  This is
+       * a vector of length size().
+       */
+      std::vector<double> initial_state() const {
+        std::vector<double> initial(size_, 0.0);
+        for (size_t i = 0; i < N_; i++)
+          initial[i] = stan::math::value_of(y0_[i]);
+        if (initial_var::value) {
+          for (size_t i = 0; i < N_; i++)
+            initial[N_ + i * N_ + i] = 1.0;
+        }
+        return initial;
+      }
+
+    private:
+      inline void
+      rhs_sens(const std::vector<double>& initial,
+               const std::vector<double>& param,
+               const std::vector<double>& z,
+               std::vector<double>& dz_dt,
+               double t) const {
+        ode_system_(t, z, dz_dt);
+      }
+
+      inline void
+      rhs_sens(const std::vector<stan::math::var>& initial,
+               const std::vector<stan::math::var>& param,
+               const std::vector<double>& z,
+               std::vector<double>& dz_dt,
+               double t) const {
+        using Eigen::Map;
+        using Eigen::MatrixXd;
+        using Eigen::VectorXd;
+
+        const std::vector<double> y_base(z.begin(), z.begin() + N_);
+        Map<VectorXd> dz_dt_eig(&dz_dt[0], N_);
+        Map<MatrixXd> dZ_dt_sens(&dz_dt[0] + N_, N_, S_);
+        Map<const MatrixXd> Z_sens(&z[0] + N_, N_, S_);
+        // write Jtheta directly into correct position of dZ_dt
+        Map<MatrixXd> dZ_dt_sens_param(&dz_dt[0] + N_ + N_ * N_, N_, M_);
+        MatrixXd Jy(N_, N_);
+
+        ode_system_.jacobian(t, y_base, dz_dt_eig, Jy, dZ_dt_sens_param);
+
+        dZ_dt_sens.leftCols(N_).setZero();
+        dZ_dt_sens += Jy * Z_sens;
+      }
+
+      inline void
+      rhs_sens(const std::vector<stan::math::var>& initial,
```

Comment:
These `rhs_sens()` functions are what are forcing this class to be in `rev` instead of in `prim`. I would suggest moving the bulk of the class back to `prim` and templated appropriately. Then have a template instantiation in `rev` like it was done before.